lxd/instance_rebuild.go

```diff
@@ -0,0 +1,240 @@
+//	"500":
+//	  $ref: "#/responses/InternalServerError"
+func instanceRebuildPost(d *Daemon, r *http.Request) response.Response {
+	s := d.State()
+
+	targetProjectName := projectParam(r)
+
+	name, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if shared.IsSnapshot(name) {
+		return response.BadRequest(fmt.Errorf("Invalid instance name"))
+	}
+
+	// Parse the request
+	req := api.InstancesPost{}
+	err = json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	instanceType, err := instancetype.New(string(req.Type))
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	var targetProject *api.Project
+	var sourceImage *api.Image
+	var sourceImageRef string
+	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		dbProject, err := dbCluster.GetProject(ctx, tx.Tx(), targetProjectName)
+		if err != nil {
+			return fmt.Errorf("Failed loading project: %w", err)
+		}
+
+		targetProject, err = dbProject.ToAPI(ctx, tx.Tx())
+		if err != nil {
+			return err
+		}
+
+		// If a remote server is being used, check whether we have a cached image for the alias.
+		// If so then use the cached image fingerprint for loading the cache image profiles.
+		// As its possible for a remote cached image to have its profiles modified after download.
+		getCachedImage := func(sourceImageRef string, sourceImageHash string) string {
+			if req.Source.Server != "" {
+				for _, architecture := range s.OS.Architectures {
+					cachedFingerprint, err := tx.GetCachedImageSourceFingerprint(ctx, req.Source.Server, req.Source.Protocol, sourceImageRef, string(req.Type), architecture)
+					if err == nil && cachedFingerprint != sourceImageHash {
+						sourceImageHash = cachedFingerprint
+						return sourceImageHash
+					}
+				}
+			}
+
+			return sourceImageHash
+		}
+
+		if req.Source.Type != "none" {
+			var sourceImageHash string
+			if req.NewSource.Alias != "" {
+				// Resolve the new image.
+				sourceImageRef, err = instance.ResolveImage(ctx, tx, targetProject.Name, req.NewSource)
+				if err != nil {
+					return err
+				}
+			} else {
+				// Resolve the current instance image.
+				sourceImageRef, err = instance.ResolveImage(ctx, tx, targetProject.Name, req.Source)
+				if err != nil {
+					return err
+				}
+			}
+
+			sourceImageHash = sourceImageRef
+			sourceImageHash = getCachedImage(sourceImageRef, sourceImageHash)
+
+			// Check if image has an entry in the database (but don't fail if not found).
+			_, sourceImage, err = tx.GetImageByFingerprintPrefix(ctx, sourceImageHash, dbCluster.ImageFilter{Project: &targetProject.Name})
+			if err != nil && !api.StatusErrorCheck(err, http.StatusNotFound) {
+				return err
+			}
+
+			// If image has an entry in the database then use its profiles if no override provided.
+			if sourceImage != nil && req.Profiles == nil {
+				req.Architecture = sourceImage.Architecture
```

Comment:
I'm not sure we should be supporting changing architecture of an instance, or replacing its profiles with the default ones from the image during rebuild.

What do you think @stgraber ?